### PR TITLE
Add Redis cache support to reduce number of requests made to Okta on token validation

### PR DESCRIPTION
**Pull request by ikas**

### CODE REVIEW

File: src/services/cache.service.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+                resolve(JSON.parse(res) as OktaUser);
+            });
+        });
+    }
+
+    static async set(key: string, value: OktaUser): Promise<any> {
+        const client: RedisClient = CacheService.getCacheClient();
+        logger.info(`[CacheService] Setting key ${key} in cache...`);
+        return new Promise(resolve => {
+            client.set(key, JSON.stringify(value), 'EX', 60 * 60 * 24, (err, res) => {
```
Comment (tiagojsag) [inline src/services/cache.service.ts]: Extract this TTL to a config value

File: src/services/cache.service.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
+import config from 'config';
+import logger from 'logger';
+import redis, {RedisClient} from 'redis';
+
+import {OktaUser} from 'services/okta.interfaces';
+
+export default class CacheService {
+    private static getCacheClient(): RedisClient {
+        const client: RedisClient = redis.createClient({ url: config.get('redis.url') as string });
```
Comment (tiagojsag) [inline src/services/cache.service.ts]: this creates a new connection to redis on each request. Is there any way we can keep that connection alive for the lifespan of the node app, so we reduce load on redis and RTT?

File: src/services/cache.service.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            });
+        });
+    }
+
+    static async clear(): Promise<any> {
+        const client: RedisClient = CacheService.getCacheClient();
+        logger.info(`[CacheService] Clearing cache...`);
+        return new Promise(resolve => {
+            client.flushall((err, res) => {
+                if (err) logger.error(err);
```
Comment (tiagojsag) [inline src/services/cache.service.ts]: on err, shouldn't we reject the promise we're returning? (same for the other functions above)

File: src/services/okta.service.ts [hunk 4]
```diff
@@ -171,33 +172,41 @@ export default class OktaService {
         // Validate token age, and only go to Okta if token is older than 1h
         const ONE_HOUR: number = 60 * 60 * 1000;
         if ((new Date().getTime() - (payload.iat * 1000)) < ONE_HOUR) {
+            logger.info('[OktaService] Token is younger than 1h, skipping validation...');
```
Comment (tiagojsag) [inline src/services/okta.service.ts:174]: 1h is hardcoded and should not be

File: src/services/cache.service.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        logger.info(`[CacheService] Setting key ${key} in cache...`);
+        return new Promise(resolve => {
+            client.set(key, JSON.stringify(value), 'EX', 60 * 60 * 24, (err, res) => {
+                if (err) { logger.error(err); }
+                resolve(res);
+            });
+        });
+    }
+
+    static async delete(user: OktaUser): Promise<void> {
```
Comment (tiagojsag) [inline src/services/cache.service.ts]: nitpick: can you rename this to "invalidate"? delete sounds too ominous

File: src/services/cache.service.ts (hunk 2 above)
Comment (ikas) [reply src/services/cache.service.ts]: should be fixed

**Pull request merged**